lib/Dialect/FIRRTL/Transforms/InferResets.cpp

```diff
@@ -0,0 +1,1461 @@
+    mlir::emitError(root.field.getValue().getLoc())
+        << "reset network never driven with concrete type";
+    return failure();
+  }
+
+  // Handle the case where we have votes for both kinds.
+  if (asyncDrives > 0 && syncDrives > 0) {
+    ResetSignal root = guessRoot(net);
+    bool majorityAsync = asyncDrives >= syncDrives;
+    auto diag =
+        mlir::emitError(root.field.getValue().getLoc())
+        << "reset network simultaneously connected to async and sync resets";
+    diag.attachNote(root.field.getValue().getLoc())
+        << "did you intend for the reset to be "
+        << (majorityAsync ? "async?" : "sync?");
+    for (auto &drive : getResetDrives(net)) {
+      if ((drive.dst.type.isa<AsyncResetType>() && !majorityAsync) ||
+          (drive.src.type.isa<AsyncResetType>() && !majorityAsync) ||
+          (drive.dst.type.isa<UIntType>() && majorityAsync) ||
+          (drive.src.type.isa<UIntType>() && majorityAsync))
+        diag.attachNote(drive.loc)
+            << "offending " << (majorityAsync ? "sync" : "async")
+            << " drive here:";
+    }
+    return failure();
+  }
+
+  // At this point we know that the type of the reset is unambiguous. If there
+  // are any votes for async, we make the reset async. Otherwise we make it
+  // sync.
+  auto kind = (asyncDrives ? ResetKind::Async : ResetKind::Sync);
+  LLVM_DEBUG(llvm::dbgs() << "- Inferred as " << kind << "\n");
+  return kind;
+}
+
+//===----------------------------------------------------------------------===//
+// Reset Updating
+//===----------------------------------------------------------------------===//
+
+LogicalResult InferResetsPass::updateReset(ResetNetwork net, ResetKind kind) {
+  LLVM_DEBUG(llvm::dbgs() << "Updating reset network with "
+                          << std::distance(net.begin(), net.end())
+                          << " nodes to " << kind << "\n");
+
+  // Determine the final type the reset should have.
+  FIRRTLType resetType;
+  if (kind == ResetKind::Async)
+    resetType = AsyncResetType::get(&getContext());
+  else
+    resetType = UIntType::get(&getContext(), 1);
+
+  // Update all those values in the network that cannot be inferred from
+  // operands. If we change the type of a module port (i.e. BlockArgument), add
+  // the module to a module worklist since we need to update its function type.
+  SmallSetVector<Operation *, 16> worklist;
+  SmallDenseSet<Operation *> moduleWorklist;
+  for (auto signal : net) {
+    Value value = signal.field.getValue();
+    if (!value.isa<BlockArgument>() &&
+        !isa_and_nonnull<WireOp, RegOp, RegResetOp, InstanceOp, InvalidValueOp>(
+            value.getDefiningOp()))
+      continue;
+    if (updateReset(signal.field, resetType)) {
+      for (auto user : value.getUsers())
+        worklist.insert(user);
+      if (auto blockArg = value.dyn_cast<BlockArgument>())
+        moduleWorklist.insert(blockArg.getOwner()->getParentOp());
+    }
+  }
+
+  // Work dat list.
```

Comment:
Please expand this comment to something like: Process the worklist of operations that have their type changed, pushing types down the SSA dataflow graph.  This is important because we change the reset types in aggregates, and then need all the subindex etc operations updated as appropriate.

Reply:
Done :+1: